### ISSUE TITLE
Strange bug when running from emacs/magit

### ISSUE THREAD

**Issue by argent-smith** [sds/overcommit#671]
Hi. I'm trying to commit from emacs/magit getting this:

```
  0 git … stash apply --index stash\@\{0\}
[Russian locale messages skipped]
  0 git … rev-parse stash\@\{0\}
38851d8333d60b04e9fb2925380980a15bb2aa56

  0 git … reflog delete --updateref --rewrite stash\@\{0\}
  0 git … update-ref -d refs/stash
  1 git … commit --
Unable to setup environment for pre-commit hook run:
STDOUT:
STDERR:error: pathspec ':/' did not match any file(s) known to git
```
Any ideas?

**Comment by sds** [comment]
I'm not sure what's causing your error, @argent-smith. All I can say is that the only way that error appears is if there is an issue creating an entry in your stash.

The error you're seeing occurs when the following command returns a non-zero exit status:
https://github.com/sds/overcommit/blob/aeb649d6f067d8cc25e61b5a338487ef998ac63b/lib/overcommit/hook_context/pre_commit.rb#L56-L58

This can happen for a few reasons. One example is if you have daemons or background processes that modify your working tree in response to changes in your working tree. For example, it automatically runs a static analysis tool that autocorrects and modifies files when errors are detected (e.g. [`rubocop ---auto-correct`](https://rubocop.readthedocs.io/en/latest/auto_correct/)). When Overcommit attempts to restore the stash, it fails because the files have been changed in a way that it doesn't expect.

This is the likely cause of the issue you're seeing. However, unless you can provide a repository where this is consistently reproducible (ideally without emacs/magit) I'm going to close this as it won't be debuggable from our side. Feel free to respond with a clearly-reproducible example and we'll be happy to take a deeper look. 

Thanks!

**Comment by argent-smith** [comment]
Thanks a lot for the analysis. I have two observations: the problem happens when I use emacs/magit on all my repos; it doesn't happen when I commit from the cli. I will send a Q to the magit team.

**Comment by argent-smith** [comment]
Magit people said this helps: https://magit.vc/manual/magit/My-Git-hooks-work-on-the-command_002dline-but-not-inside-Magit.html

Well, how to unset GIT_LITERAL_PATHSPECS?

**Comment by argent-smith** [comment]
```

**Comment by sds** [comment]
Hey @argent-smith, unfortunately setting the `env` option will only set environment variables for each individual hook, which happens separately from when the stash commands are executed. See https://github.com/sds/overcommit/blob/aeb649d6f067d8cc25e61b5a338487ef998ac63b/lib/overcommit/hook/base.rb#L47

I believe we can solve this by making the following change:

**fix.patch**
```diff
:100644 100644 2378de5b 00000000 M	lib/overcommit/hook_context/pre_commit.rb

diff --git a/lib/overcommit/hook_context/pre_commit.rb b/lib/overcommit/hook_context/pre_commit.rb
index 2378de5b..aa02120f 100644
--- a/lib/overcommit/hook_context/pre_commit.rb
+++ b/lib/overcommit/hook_context/pre_commit.rb
@@ -54,7 +54,7 @@ module Overcommit::HookContext
 
         stash_message = "Overcommit: Stash of repo state before hook run at #{Time.now}"
         result = Overcommit::Utils.execute(
-          %w[git -c commit.gpgsign=false stash save --keep-index --quiet] + [stash_message]
+          %w[git -c commit.gpgsign=false --literal-pathspecs stash save --keep-index --quiet] + [stash_message]
         )
 
         unless result.success?

```

However, since I don't have a way to confirm this locally (I have no way to reproduce), I would appreciate it if you could test for us.

In the repository where you have set up Overcommit, you can run:

```
patch -p1 -d "$(bundle show overcommit)" < fix.patch
```

...where the contents of `fix.patch` are the patch above. Try running your command, and if you don't see the error, it's fixed!

You can undo the patch by running:
```
patch -R -p1 -d "$(bundle show overcommit)" < fix.patch
```

Run it one more time to make sure the error is re-introduced. 

Once we've confirmed this actually works, we can make a PR to change the behavior. Appreciate your help!

**Comment by argent-smith** [comment]
@sds great! I will. Will report later today. Thnx!

**Comment by argent-smith** [comment]
@sds actually, we need GIT_LITERAL_PATHSPECS to be set to 0 (cause 1 introduced by magit breaks the thing). So, this works:

```fix-ag.patch
:100644 100644 2378de5b 00000000 M	lib/overcommit/hook_context/pre_commit.rb

diff --git a/lib/overcommit/hook_context/pre_commit.rb b/lib/overcommit/hook_context/pre_commit.rb
index 2378de5b..aa02120f 100644
--- a/lib/overcommit/hook_context/pre_commit.rb
+++ b/lib/overcommit/hook_context/pre_commit.rb
@@ -54,5 +54,5 @@ module Overcommit::HookContext

         stash_message = "Overcommit: Stash of repo state before hook run at #{Time.now}"
         result = Overcommit::Utils.execute(
-          %w[git -c commit.gpgsign=false stash save --keep-index --quiet] + [stash_message]
+          %w[env GIT_LITERAL_PATHSPECS=0 git -c commit.gpgsign=false stash save --keep-index --quiet] + [stash_message]
         )

         unless result.success?
```

**Comment by sds** [comment]
Released [0.50.0](https://github.com/sds/overcommit/releases/tag/v0.50.0) with the fix.

**Comment by argent-smith** [comment]
❤️
